From 4443fcd38fe1682556e1851df97c7d652d078f3d Mon Sep 17 00:00:00 2001
From: Tim Foster <tim.foster@joyent.com>
Date: Mon, 28 Jan 2019 11:57:48 +0000
Subject: [PATCH] TOOLS-2163 eng.git 'check' target fails after TOOLS-2062
 TOOLS-2164 bits-upload.sh should look for .tar.gz files for agents TOOLS-2165
 bits-upload.sh should bail out when stamp ends with '-dirty' TOOLS-2166 'make
 show-buildenv' should emit the recommended jenkins-agent image uuid Reviewed
 by: Chris Burroughs <chris.burroughs@joyent.com> Approved by: Chris Burroughs
 <chris.burroughs@joyent.com>

---
 deps/eng                   |  1 +
 tools/bits-upload.sh       | 26 +++++++++++++++++++++++++-
 tools/validate-buildenv.sh | 20 +++++++++++++++++++-
 3 files changed, 45 insertions(+), 2 deletions(-)
 create mode 120000 deps/eng

diff --git a/deps/eng b/deps/eng
new file mode 120000
index 0000000..a96aa0e
--- /dev/null
+++ b/deps/eng
@@ -0,0 +1 @@
+..
\ No newline at end of file
diff --git a/tools/bits-upload.sh b/tools/bits-upload.sh
index 50745de..86f46d2 100755
--- a/tools/bits-upload.sh
+++ b/tools/bits-upload.sh
@@ -45,6 +45,11 @@ set -o errexit
 #
 BITS_UPLOAD_OVERWRITE=false
 
+#
+# Whether we should allow upload of bits marked with a '-dirty' $STAMP
+#
+BITS_UPLOAD_ALLOW_DIRTY=$ENGBLD_BITS_UPLOAD_ALLOW_DIRTY
+
 #
 # A path to our updates-imgadm command
 #
@@ -301,13 +306,17 @@ function publish_to_updates {
 
         # Some payloads are not zfs-based, look for likely alternatives.
         # This assumes that a single directory with <file>.manifest
-        # contains only one of [<file>.zfs.gz, <file>.sh, <file>.tgz]
+        # contains only one of [<file>.zfs.gz, <file>.sh, <file>.tgz,
+	# <file>.tar.gz]
         if [[ ! -f "${IMAGEFILE}" ]]; then
             IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.sh/g')
         fi
         if [[ ! -f "${IMAGEFILE}" ]]; then
             IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.tgz/g')
         fi
+        if [[ ! -f "${IMAGEFILE}" ]]; then
+            IMAGEFILE=$(echo ${MF} | sed -e 's/\..*manifest$/.tar.gz/g')
+        fi
 
         if [[ ! -f ${IMAGEFILE} ]]; then
             echo "Unable to determine image file for ${MF}."
@@ -444,6 +453,21 @@ if [[ -z "$TIMESTAMP" ]]; then
     fi
 fi
 
+#
+# Attempting to upload bits from a workspace marked dirty isn't
+# allowed, because then we have no sure way to get from the development
+# bits to the corresponding source commit. Arguably, this _might_ be
+# feasible for pure-javascript-only components. Developers who absolutely
+# must upload bits marked '-dirty' can always upload them manually, or
+# use the BITS_UPLOAD_ALLOW_DIRTY escape hatch in case of emergency.
+#
+set +o errexit
+echo $TIMESTAMP | grep -q '\-dirty$'
+if [[ $? -eq 0 && -z "$BITS_UPLOAD_ALLOW_DIRTY" ]]; then
+    fatal "Bits timestamp $TIMESTAMP marked 'dirty': not uploading"
+fi
+set -o errexit
+
 UPLOAD_SUBDIR=$TIMESTAMP
 
 if [[ -n "$USE_LOCAL" ]]; then
diff --git a/tools/validate-buildenv.sh b/tools/validate-buildenv.sh
index 9241eb2..27adefc 100755
--- a/tools/validate-buildenv.sh
+++ b/tools/validate-buildenv.sh
@@ -430,7 +430,7 @@ function validate_build_platform {
 #
 function print_required_pkgsrc_version {
     if [[ -n "$REQUIRED_IMAGE" ]]; then
-        echo "${PKGSRC_MAP[$REQUIRED_IMAGE]} ${SDC_MAP[$REQUIRED_IMAGE]}"
+        echo "${PKGSRC_MAP[$REQUIRED_IMAGE]} ${SDC_MAP[$REQUIRED_IMAGE]} ${JENKINS_AGENT_MAP[$REQUIRED_IMAGE]}"
         exit 0
     else
         exit 1
@@ -612,6 +612,20 @@ function validate_non_pkgsrc_bins {
     return 0
 }
 
+#
+# Issue a warning to the developer if their workspace contains uncommitted
+# changes, which would result in bits-upload.sh not posting any built bits
+# to Manta or updates.joyent.com
+#
+function verify_clean_repo {
+    HAS_DIRTY=$(git describe --all --long --dirty | grep '\-dirty$')
+    if [[ -n "$HAS_DIRTY" ]]; then
+        echo "WARNING: this workspace contains uncommitted changes,"
+        echo "which means that any build artifacts will not be uploaded by"
+        echo "bits-upload.sh to either Manta or updates.joyent.com"
+    fi
+}
+
 function usage {
     echo "Usage: validate-build-platform [-h] [-r]"
     echo "  -h       print usage"
@@ -660,6 +674,10 @@ else
     RESULT=$(( $RESULT + $? ))
     validate_non_pkgsrc_bins
     RESULT=$(( $RESULT + $? ))
+    # this doesn't contribute to success/failure, but warns
+    # developers that '-dirty' repositories will result in
+    # bits-upload not posting to Manta/updates.joyent.com.
+    verify_clean_repo
     if [[ "$RESULT" -gt 0 ]]; then
         exit 1
     else
-- 
2.21.0

